mmgen/models/diffusions/base_diffusion.py

```diff
@@ -0,0 +1,903 @@
+
+        return outputs
+
+    @torch.no_grad()
+    def DDPM_sample(self,
+                    model,
+                    noise=None,
+                    num_batches=0,
+                    label=None,
+                    save_intermedia=False,
+                    timesteps_noise=None,
+                    **kwargs):
+        """DDPM sample from random noise.
+        Args:
+            model (torch.nn.Module): Denoising model used to sample images.
+            noise (torch.Tensor | callable | None): You can directly give a
+                batch of noise through a ``torch.Tensor`` or offer a callable
+                function to sample a batch of noise data. Otherwise, the
+                ``None`` indicates to use the default noise sampler.
+            num_batches (int, optional):
+            label (torch.Tensor, optional):
+            save_intermedia (bool, optional): Whether to save denoising result
+                of intermedia timesteps. If set as True, would return a list
+                contains denoising results of each timestep. Otherwise, only
+                the final denoising result would be returned. Defaults to
+                False.
+            timesteps_noise (torch.Tensor, optional): Noise term used in each
+                denoising timesteps. If passed, the input noise would shape as
+                [num_timesteps, b, c, h, w]. If set as None, noise of each
+                denoising timestep would be randomly sampled. Default as None.
+        Returns:
+            torch.Tensor | list[torch.Tensor]: If ``save_intermedia``, a list
+                contains denoising results of each timestep would be returned.
+                Otherwise, only the final denoising result would be returned.
+        """
+        device = get_module_device(self)
+        x_t = self.get_noise(noise, num_batches=num_batches).to(device)
+        if save_intermedia:
+            # save input
+            intermedia = [x_t.clone()]
+
+        # use timesteps noise if defined
+        if timesteps_noise is not None:
+            timesteps_noise = self.get_noise(
+                timesteps_noise, num_batches=num_batches,
+                timesteps_noise=True).to(device)
+
+        batched_timesteps = torch.arange(self.num_timesteps - 1, -1,
+                                         -1).long().to(device)
+        for t in batched_timesteps:
+            batched_t = t.expand(x_t.shape[0])
+            step_noise = timesteps_noise[t, ...] \
+                if timesteps_noise is not None else None
+
+            x_t = self.denoising_step(
+                model, x_t, batched_t, noise=step_noise, label=label, **kwargs)
+            if save_intermedia:
+                intermedia.append(x_t.clone())
+        if save_intermedia:
+            return intermedia
+        return x_t
+
+    def prepare_diffusion_vars(self):
+        """Prepare for variables used in the diffusion process.
+
+        TODO: we should use cupy or torch for speeding up?
```

Comment:
If it is convenient, we can directly use torch. The current version with numpy is also ok for me.